gax/src/main/java/com/google/api/gax/batching/v2/BatchingSettings.java

```diff
@@ -0,0 +1,120 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching.v2;
+
+import com.google.api.core.BetaApi;
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+
+/**
+ * Represents the batching settings to use for an API method that is capable of batching.
+ *
+ * <p>Each batching client must define their own set of default values for these thresholds, which
+ * would be the safest behavior for their jobs.
+ *
+ * <p>The default instance of this settings are configured to accept elements until either of the
+ * threshold reaches {@link Long#MAX_VALUE}) or an explicit call to {@link Batcher#flush()} is made
+ * or the {@link Batcher#close()} is called. Users are expected to configure actual batching
+ * thresholds explicitly: the element count or the request bytes count.
+ *
+ * <p>Warning: With the incorrect settings, it is possible to cause long periods of dead waiting
+ * time.
+ *
+ * <p>When batching is configured for an API method, a call to that method will result in the
+ * request being queued up with other requests. When any of the set thresholds are reached, the
+ * queued up requests are packaged together in a batch and set to the service as a single RPC. When
+ * the response comes back, it is split apart into individual responses according to the individual
+ * input requests.
+ *
+ * <p>There are several supported thresholds:
+ *
+ * <ul>
+ *   <li><b>Message Count Threshold</b>: Once this many messages are queued, send all of the
+ *       messages in a single call, even if the request byte threshold has not been exceed yet. The
+ *       default value is {@link Long#MAX_VALUE} messages.
+ *   <li><b>Request Byte Threshold</b>: Once the number of bytes in the batched request reaches this
+ *       threshold, send all of the messages in a single call, even if message count threshold has
+ *       not been exceeded yet. The default value is {@link Long#MAX_VALUE} bytes.
+ * </ul>
+ *
+ * <p>These thresholds are treated as triggers, not as limits. Each threshold is an independent
+ * trigger and doesn't have any knowledge of the other thresholds.
+ */
+@BetaApi("The surface for batching is not stable yet and may change in the future.")
+@AutoValue
+public abstract class BatchingSettings {
+
+  /** Get the element count threshold to use for batching. */
+  public abstract long getElementCountThreshold();
+
+  /** Get the request byte threshold to use for batching. */
+  public abstract long getRequestByteThreshold();
+
+  /** Get a new builder. */
+  public static Builder newBuilder() {
+    return new AutoValue_BatchingSettings.Builder()
+        .setElementCountThreshold(Long.MAX_VALUE)
+        .setRequestByteThreshold(Long.MAX_VALUE);
```

Comment:
Maybe use some reasonably big value here (and in element count) instead of `MAX_VALUE`? People most probably will go with default values first (yes, everything is documented, but people read documentaiton only after something broke..). The default values can result in `OutOfMemoryException` quite easily.

Reply:
There are 2 audiences here:
- client developers
- end users

The client developers will configure defaults per service in their StubSettings. Each service will have different profiles. When looking at the code:
```java
  BatchingSettings.newBuilder().setMaxElementCount(100).build(); 
```
I would be very surprised if it flushed a single large element. I would want it to be spelled out in the StubSetting defaults as 

```java
  BatchingSettings.newBuilder().setElementCountThreshold(100).setRequestByteThreshold(1024).build(); 
```


The other audience is the end users, the expectations with them is that they should use toBuilder() on the client developer provided instance to tweak stuff

Reply:
@igorbernstein2 Who are the client developers? Are  you talking about the manual layer on top of this? If yes, then most of the generated clients don't have that wrapper, so the there is simply no `toBuilder()` provided for them by the client developer.

Reply:
I'm ok with no default